scripts/dev/tidy.php

```diff
@@ -0,0 +1,394 @@
+    'config.sub',
+
+    // Visual Studio generated templates
+    'win32/php7dllts.rc',
+    'win32/php7ts.rc',
+    'win32/php7ts_cli.rc',
+];
+
+// List of common ignored regular expressions
+$ignoredPatterns = [
+    // Known binary files based on extension
+    '/^.+\.tar\.gz$/',
+    '/^.+\.png$/',
+    '/^.+\.jpeg$/',
+    '/^.+\.jpg$/',
+    '/^.+\.tiff$/',
+    '/^.+\.phar$/',
+    '/^.+\.bz2$/',
+    '/^.+\.zip$/',
+    '/^.+\.data$/',
+    '/^.+\.webp$/',
+    '/^.+\.bmp$/',
+    '/^.+\.ttf$/',
+    '/^.+\.gd2$/',
+    '/^.+\.gif$/',
+    '/^.+\.ico$/',
+    '/^.+\.mov$/',
+    '/^.+\.gz$/',
+    '/^.+\.patch$/',
+    '/^.+\~$/',
+];
+
+// PHP specific ignored regular expressions
+$phpIgnoredPatterns = [
+    // Extensions and SAPIs tests
+    '/^sapi\/[a-zA-Z0-9_-]+\/tests\/.+$/',
+    '/^ext\/[a-zA-Z0-9_-]+\/tests\/.+$/',
+];
+
+// Guess if provided $path is the root of the php-src repository
+if (array_filter([$path.'/build', $path.'/ext', $path.'/main/php.h', $path.'/Zend'], 'file_exists')) {
+    $ignores = array_merge($ignores, $phpIgnores);
+    $ignoredPatterns = array_merge($ignoredPatterns, $phpIgnoredPatterns);
+}
+
+/**
+ * RecursiveCallbackFilterIterator filter for retrieving files.
+ */
+$filter = function ($file, $key, $iterator) use ($ignores, $ignoredPatterns, $path) {
+    $pathnameWithoutBase = str_replace($path.'/', '', $file->getPathname());
+
+    foreach ($ignoredPatterns as $regex) {
+        if (preg_match($regex, $pathnameWithoutBase)) {
+            return false;
+        }
+    }
+
+    return (!in_array($pathnameWithoutBase, $ignores));
+};
+
+/**
+ * Return message colored in green.
+ */
+$info = function (string $message) use ($colors, $quiet) {
+    if ($quiet) {
+        return null;
+    }
+
+    $green = ($colors) ? "\e[32m" : '';
+    $default = ($colors) ? "\e[39m" : '';
+
+    echo $green.$message.$default."\n";
+};
+
+/**
+ * Return message colored in red.
+ */
+$error = function (string $message) use ($colors, $quiet) {
+    if ($quiet) {
+        return null;
+    }
+
+    $red = ($colors) ? "\e[31m" : '';
+    $default = ($colors) ? "\e[39m" : '';
+
+    echo $red.$message.$default."\n";
+};
+
+$files = [];
+
+if (is_file($path)) {
+    $files = [$path];
+} else {
+    $innerIterator = new \RecursiveDirectoryIterator(
+        $path,
+        \RecursiveDirectoryIterator::SKIP_DOTS
+    );
+
+    $iterator = new \RecursiveIteratorIterator(
+        new \RecursiveCallbackFilterIterator($innerIterator, $filter)
+    );
+
+    foreach ($iterator as $pathName => $fileInfo) {
+        $files[] = $pathName;
+    }
+}
+
+$info(($fix ? 'Tidying' : 'Checking')." trailing whitespace and final new lines:");
+
+$exitCode = 0;
+
+foreach ($files as $file) {
+    $content = file_get_contents($file);
+
+    // Clean whitespace
+    $trimmedWhitespace = ($whitespace) ? trimTrailingWhitespace($content) : $content;
+
+    if ($trimmedWhitespace !== $content) {
+        $error($file.': has whitespace');
```

Comment:
There are a number of places here where repeated conditions can be extracted, e.g.:

```
if ($trimmedWhitespace !== $content) {
    if ($fix) { ... }
    else { ... }
}
```

Actually that's not the same as the current code seems to be printing both the error and the info in fix mode -- is that intentional?

Reply:
I've simplified this one more and moved those into the final if statement where saving happens. Yes, now there are still multiple infos when saving. I'll recheck it out.